src/chips/sam4l/dma.rs

```diff
@@ -265,87 +265,33 @@ impl DMAChannel {
     }
 }
 
-#[no_mangle]
-#[allow(non_snake_case)]
-pub unsafe extern fn PDCA_0_Handler() {
-    use common::Queue;
-    use nvic;
-    use chip;
-
-    let registers : &mut DMARegisters =
-        mem::transmute(DMAChannels[0].registers);
-    volatile_store(&mut registers.interrupt_disable, 0xffffffff);
-    nvic::disable(nvic::NvicIdx::PDCA0);
-    chip::INTERRUPT_QUEUE.as_mut().unwrap().enqueue(nvic::NvicIdx::PDCA0);
-}
-
-#[no_mangle]
-#[allow(non_snake_case)]
-pub unsafe extern fn PDCA_1_Handler() {
-    use common::Queue;
-    use nvic;
-    use chip;
-
-    let registers : &mut DMARegisters =
-        mem::transmute(DMAChannels[1].registers);
-    volatile_store(&mut registers.interrupt_disable, 0xffffffff);
-    nvic::disable(nvic::NvicIdx::PDCA1);
-    chip::INTERRUPT_QUEUE.as_mut().unwrap().enqueue(nvic::NvicIdx::PDCA1);
-}
-
-#[no_mangle]
-#[allow(non_snake_case)]
-pub unsafe extern fn PDCA_2_Handler() {
-    use common::Queue;
-    use nvic;
-    use chip;
-
-    let registers : &mut DMARegisters =
-        mem::transmute(DMAChannels[2].registers);
-    volatile_store(&mut registers.interrupt_disable, 0xffffffff);
-    nvic::disable(nvic::NvicIdx::PDCA2);
-    chip::INTERRUPT_QUEUE.as_mut().unwrap().enqueue(nvic::NvicIdx::PDCA2);
-}
-
-#[no_mangle]
-#[allow(non_snake_case)]
-pub unsafe extern fn PDCA_3_Handler() {
-    use common::Queue;
-    use nvic;
-    use chip;
-
-    let registers : &mut DMARegisters =
-        mem::transmute(DMAChannels[3].registers);
-    volatile_store(&mut registers.interrupt_disable, 0xffffffff);
-    nvic::disable(nvic::NvicIdx::PDCA3);
-    chip::INTERRUPT_QUEUE.as_mut().unwrap().enqueue(nvic::NvicIdx::PDCA3);
-}
-
-#[no_mangle]
-#[allow(non_snake_case)]
-pub unsafe extern fn PDCA_4_Handler() {
-    use common::Queue;
-    use nvic;
-    use chip;
-
-    let registers : &mut DMARegisters =
-        mem::transmute(DMAChannels[4].registers);
-    volatile_store(&mut registers.interrupt_disable, 0xffffffff);
-    nvic::disable(nvic::NvicIdx::PDCA4);
-    chip::INTERRUPT_QUEUE.as_mut().unwrap().enqueue(nvic::NvicIdx::PDCA4);
+macro_rules! pdca_handler {
+    ($name: ident, $nvic: ident, $num: expr) => {
+        interrupt_handler!(
```

Comment:
I think this is a good set of changes and removes a lot of duplicated code, but I do always start to get a little leery when you have nested `#define`s, or in this case templates. Can make code hard to read. I think this is probably fine, but I wouldn't call pdca_handler in another template :)

Does rust have an equivalent to `gcc -E`, that is a compile step that expands all templates into the resulting rust code? 30 seconds of googling failed to find it.